### PR TITLE
docker_perf: initial commit

### PR DESCRIPTION
**Pull request by pietrushnic**
Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

### CODE REVIEW

File: _posts/2018-08-22-optimize-performance-in-docker.md [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+(...)
+[00:05:50.237480] Successfully tagged 3mdeb/xen-docker:latest
+```
+
+Assuming that the network conditions did not change between runs to extent of
+30s delay we can conclude:
+
+* using cacher even with cold cache is better than nothing, it gives the
+  speedup of about 5%.
+* using filled cache can give ~20% over container build time, if significant
```
Comment (macpijan) [inline _posts/2018-08-22-optimize-performance-in-docker.md]: @pietrushnic The `filled` cache concept was not mentioned before

File: _posts/2018-08-22-optimize-performance-in-docker.md [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+
+No `ccache` single-threaded:
+
+```
+docker run --rm -it -v $PWD:/home/xen -w /home/xen 3mdeb/xen-docker make debball| ts -s '[%.T]'
+(...)
+[00:13:10.006206] dpkg-deb: building package 'xen-upstream' in 'xen-upstream-4.8.4.deb'.
+```
+
+Cold cache:
```
[the remark this answers is not in the document]
Comment (pietrushnic) [reply _posts/2018-08-22-optimize-performance-in-docker.md]: @macpijan fixed.

File: _posts/2018-08-22-optimize-performance-in-docker.md (hunk 1 above)
Comment (pietrushnic) [reply _posts/2018-08-22-optimize-performance-in-docker.md]: @macpijan I assumed this is obvious :) Of course I meant `hot` instead of `filled`.

**Comment by pietrushnic** [comment]
@macpijan thanks for fixes. @miczyg1 @macpijan @mscibisz I think this post is ready to publish. Please let me know if there are any problems with it.

**Comment by mscibisz** [comment]
@miczyg1 @macpijan Do you think it is ready to be published?

File: _posts/2018-08-22-optimize-performance-in-docker.md [hunk 3]
```diff
@@ -0,0 +1,487 @@
...
+```
+
+We should also see that cacher listens on port `3142`:
+
+```
+[16:40:01] pietrushnic:~ $ netstat -an |grep 3142
+tcp6       0      0 :::3142                 :::*                    LISTEN
+```
+
+Dockerfile should contain following environment variable:
```
Comment (miczyg1) [inline _posts/2018-08-22-optimize-performance-in-docker.md]: This context switch is too sudden, it is hard to distinguish if it still relates to apt-cacher Dockerfile or target xen-docker Dockerfile. Maybe start sentence with: `Now our target xen-docker Dockerfile should contain ...`.

File: _posts/2018-08-22-optimize-performance-in-docker.md [hunk 4]
```diff
@@ -0,0 +1,487 @@
...
+ENV http_proxy ${http_proxy}
+```
+
+Now we can run docker building with appropriate parameters:
+
+```
+docker build --build-arg http_proxy=http://<CACHER_IP>:3142/ -t 3mdeb/xen-docker .| ts -s '[%.T]'
+```
+
+## performance measures
```
Comment (miczyg1) [inline _posts/2018-08-22-optimize-performance-in-docker.md]: Capital letter?

File: _posts/2018-08-22-optimize-performance-in-docker.md [hunk 5]
```diff
@@ -0,0 +1,487 @@
...
+(...)
+[00:05:50.237480] Successfully tagged 3mdeb/xen-docker:latest
+```
+
+Assuming that the network conditions did not change between runs to extent of
+30s delay we can conclude:
+
+* using cacher even with cold cache is better than nothing, it gives the
+  speedup of about 5%
+* using hot cache can give ~20% over container build time, if significant
```
Comment (miczyg1) [inline _posts/2018-08-22-optimize-performance-in-docker.md]: `can spare ~20% of normal container build time`

File: _posts/2018-08-22-optimize-performance-in-docker.md [hunk 6]
```diff
@@ -0,0 +1,487 @@
...
+  amount of that time is package installation
+
+Of course, those numbers should be confirmed statistically.
+
+# Let's try something more complex
+
+Finally we can try to run much more sophisticated stuff like our
+[debian-rootfs-builder](https://github.com/3mdeb/debian-rootfs-builder). This
+code contain mostly compilatin and package installation through `apt-get`.
```
Comment (miczyg1) [inline _posts/2018-08-22-optimize-performance-in-docker.md]: `compilation`

File: _posts/2018-08-22-optimize-performance-in-docker.md [hunk 7]
```diff
@@ -0,0 +1,487 @@
...
+linux-kernel : make mrproper -------------------------------------------- 6.55s
+Gathering Facts --------------------------------------------------------- 4.81s
+linux-kernel : remove everything except artifacts ----------------------- 3.40s
+linux-kernel : remove everything except artifacts ----------------------- 3.11s
+linux-kernel : make olddefconfig ---------------------------------------- 3.08s
+Gathering Facts --------------------------------------------------------- 2.74s
+Playbook run took 0 days, 0 hours, 23 minutes, 19 seconds
+```
+
+This is not significant but we fain another 13% and now build time is
```
Comment (miczyg1) [inline _posts/2018-08-22-optimize-performance-in-docker.md]: fain ??

File: _posts/2018-08-22-optimize-performance-in-docker.md [hunk 8]
```diff
@@ -0,0 +1,487 @@
...
+reasonable. Still most time consuming tasks belong to compilation and package
```
Comment (miczyg1) [inline _posts/2018-08-22-optimize-performance-in-docker.md]: `time-consuming`

File: _posts/2018-08-22-optimize-performance-in-docker.md [hunk 9]
```diff
@@ -0,0 +1,487 @@
...
+	- Firmware
+	- Miscellaneous
+
+---
+
+In 3mdeb we use Docker heavily. Main tasks that we perform using it are:
+
+* firmware and embedded software building - each software in Embedded System
+  requires little bit different building environment, configuring those
+  development environments on your host quickly made a mess in your system for
```
[the remark this answers is not in the document]
Comment (mscibisz) [reply _posts/2018-08-22-optimize-performance-in-docker.md]: `may quickly make a mess` sounds better to me

File: _posts/2018-08-22-optimize-performance-in-docker.md (hunk 7 above)
Comment (mscibisz) [reply _posts/2018-08-22-optimize-performance-in-docker.md]: `gain` I guess

File: _posts/2018-08-22-optimize-performance-in-docker.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (pietrushnic) [reply _posts/2018-08-22-optimize-performance-in-docker.md]: fixed

File: _posts/2018-08-22-optimize-performance-in-docker.md [hunk 10]
```diff
@@ -0,0 +1,487 @@
...
+* [zephyr-docker](https://github.com/3mdeb/zephyr-docker) - container for
+  building Zephyr RTOS
+* [esp-open-sdk-docker](https://github.com/3mdeb/esp-open-sdk-docker) -
+  container for bulding ESP Open SDK Espressify Open RTOS
+* [docker-mbed-cli](https://github.com/3mdeb/docker-mbed-cli) - container for
+  building and running CLI of mbedOS
+* [arch-pkg-builder](https://github.com/3mdeb/arch-pkg-builder) - container for
+  building Arch packages
+
+Some are actively maintained some are not, some came from other projects, some
```
[the remark this answers is not in the document]
Comment (pietrushnic) [reply _posts/2018-08-22-optimize-performance-in-docker.md]: fixed

File: _posts/2018-08-22-optimize-performance-in-docker.md [hunk 11]
```diff
@@ -0,0 +1,487 @@
...
+where created from scratch, but all of them have value for Embedded Systems
```
[the remark this answers is not in the document]
Comment (pietrushnic) [reply _posts/2018-08-22-optimize-performance-in-docker.md]: fixed

File: _posts/2018-08-22-optimize-performance-in-docker.md [hunk 12]
```diff
@@ -0,0 +1,487 @@
...
+  caching proxy like `apt-cacher-ng`, we will show how to use it with Docker on
+  build and runtime
+
+# ccache
+
+Following example will show `ccache` usage with `xen-docker`. Great post about
+that topic was published by Tim Potter [here](http://frungy.org/docker/using-ccache-with-docker).
+
+Of course, to use `ccache` in our container we need it installed, so make sure
+your Docker file contains that package. You can take a look at [xen-docker Dockerfile](https://github.com/3mdeb/xen-docker/blob/master/Dockerfile#L15).
```
[the remark this answers is not in the document]
Comment (pietrushnic) [reply _posts/2018-08-22-optimize-performance-in-docker.md]: fixed

File: _posts/2018-08-22-optimize-performance-in-docker.md [hunk 13]
```diff
@@ -0,0 +1,487 @@
...
+cache hit (preprocessed)               0
+cache miss                             0
+cache hit rate                      0.00 %
+cleanups performed                     0
+files in cache                         0
+cache size                           0.0 kB
+max cache size                       5.0 GB
+```
+
+More to that it is important to pay attention to directory structure and
```
[the remark this answers is not in the document]
Comment (pietrushnic) [reply _posts/2018-08-22-optimize-performance-in-docker.md]: fixed

File: _posts/2018-08-22-optimize-performance-in-docker.md (hunk 4 above)
Comment (pietrushnic) [reply _posts/2018-08-22-optimize-performance-in-docker.md]: fixed

File: _posts/2018-08-22-optimize-performance-in-docker.md (hunk 5 above)
Comment (pietrushnic) [reply _posts/2018-08-22-optimize-performance-in-docker.md]: fixed

File: _posts/2018-08-22-optimize-performance-in-docker.md (hunk 6 above)
Comment (pietrushnic) [reply _posts/2018-08-22-optimize-performance-in-docker.md]: fixed

File: _posts/2018-08-22-optimize-performance-in-docker.md (hunk 7 above)
Comment (pietrushnic) [reply _posts/2018-08-22-optimize-performance-in-docker.md]: fixed

File: _posts/2018-08-22-optimize-performance-in-docker.md (hunk 8 above)
Comment (pietrushnic) [reply _posts/2018-08-22-optimize-performance-in-docker.md]: fixed

**Comment by pietrushnic** [comment]
@miczyg1 let me know if this blog post is fine for you

File: _posts/2018-08-22-optimize-performance-in-docker.md [hunk 14]
```diff
@@ -0,0 +1,487 @@
...
+RUN mkdir /home/xen/.ccache && \
+	chown xen:xen /home/xen/.ccache
+```
+
+Then to run container with `ccache` we can pass our `~/.ccache` as volume. For
+single-threaded compilation assuming you checked out correct code and called
+`./configure`:
+
+```
+
```
[the remark this answers is not in the document]
Comment (miczyg1) [reply _posts/2018-08-22-optimize-performance-in-docker.md]: @pietrushnic ?

File: _posts/2018-08-22-optimize-performance-in-docker.md [hunk 15]
```diff
@@ -0,0 +1,487 @@
...
+To conclude, we can gain even 30% with hot cache. Biggest gain we have when
+using multithreading, but this highly depends on CPU, in my case I had 8 jobs
+run simultaneously and gain was 40% in compilation time.
+
+# apt-cacher-ng
+
+There 2 use case for `apt-cacher-ng` in our workflows. One is Docker build
+time, which can be time-consuming since all packages and its dependencies are
+installed in the base image. Second is runtime, when you need some package that
+may have extensive dependencies e.g. `xen-systema-amd64`.
```
[the remark this answers is not in the document]
Comment (miczyg1) [reply _posts/2018-08-22-optimize-performance-in-docker.md]: @pietrushnic

**Comment by pietrushnic** [comment]
@Moviss please merge and publish that blog post

**Pull request merged**